Add stacktrace colors to showing `MethodError`s and method lists

Seelengrab opened the pull request:
This basically takes the module coloring logic from printing stacktraces and applies it to method error printing as well. Also contains some formatting to make the "closest candidates" stand out more by having an empty line precede & follow that block. 

<details>
<summary>Before (master) </summary>

![screenshot_2022-06-09_19-51-23_027029655](https://user-images.githubusercontent.com/11753998/172912291-15772fe5-372f-4581-bca0-fdb9427d84ff.png)

</details>


<details>
<summary>After (This PR) </summary>

![after1](https://user-images.githubusercontent.com/11753998/172908375-04b3d455-b4db-4d08-b2f3-32078312ce46.png)
![after2](https://user-images.githubusercontent.com/11753998/172908367-570b5967-3105-4932-bcc8-2d14c250af04.png)


</details>

<details>
<summary>Code for producing the above </summary>

example.jl
```julia
f() = 1
f(x,y,z; f=true) = 3
f(x,y,z,k; f=false) = 4

include("subdir/subdir2/module.jl")

h() = f(1)
l() = h()

ambig(x, y) = 1
ambig(x::Integer, y) = 2
ambig(x, y::Integer) = 3
ambig(x::Int, y::Int) = 4
ambig(x::Number, y) = 5
```

subdir/subdir2/module.jl 
```julia
module Example
    Main.f(x,y) = 2
    g() = :g
end
```


</details>

~The one thing that kind of bothers me though is that the module coloring between the `MethodError` message and the followed stacktrace is not consistent - they're done seperately & in isolation and there's no easy way to "pass on" the existing coloring that's built in `MethodError`.~ The colors between method printing, stacktraces and `MethodError` are now also consistent. There may be a race condition on the underlying `IdDict`, should two tasks set a color at the same time. Since only one printing call would be affected and it's a non-functional part, I'd say this is fine. 

Opinions, improvement ideas etc. are welcome! I've been bothered by how hard `MethodError`s are to read for a long time now, and I think unifying the printing with stacktraces is a good first step to improve this.

Closes #40251 Closes https://github.com/JuliaLang/julia/issues/40913 Closes https://github.com/JuliaLang/julia/issues/30110 Closes https://github.com/JuliaLang/julia/issues/36129

Seelengrab committed:
Add color pass to MethodError showing, like in stacktraces

Fix error printing issues

Incorporate method list printing

Fix almost all test cases

Remove complete inner module printing from tests

This is to be consistent with stacktrace printing, which now shows the topmost module
that isn't `Main`, to make it easier to identify packages in stacktraces.

Contract home dir in method printing as well

Fix now contracted user dir paths

Fix failing ambiguity test cases as well as other failing test cases

Fix missing newline in ambiguity error

Seelengrab commented:
I do expect tests to fail, mostly due to doctests. I just wanted to have CI churn away at finding them instead of my laptop :sweat_smile: I'll fix them as they occur or after some feedback about the general layout/design.

rikhuijzer commented on base/errorshow.jl:
```diff
@@ -543,7 +548,7 @@ function show_method_candidates(io::IO, ex::MethodError, @nospecialize kwargs=()
     if !isempty(lines) # Display up to three closest candidates
         Base.with_output_color(:normal, io) do io
-            print(io, "\nClosest candidates are:")
+            print(io, "\n\nClosest candidates are:")
```
```suggestion
            print(io, "\nClosest candidates are:")
```
Fewer lines is better IMO. Especially with long stacktraces.

Seelengrab replied:
This was expressly added to seperate the blocks visually and to make it easier to read - removing them is IMO very counter to what I'm trying to achieve here.

Seelengrab replied:
For comparison, here's what the current PR would look like with those newlines removed:

<details>
<summary> Click for screenshot </summary>

![screenshot_2022-04-24_15-44-31_570294865](https://user-images.githubusercontent.com/11753998/164979531-460b89b9-bdb0-400a-883b-8059ac1c6b1b.png)

</details>

IMO that is harder to read, because it's all one big blob of text, even with the color change.

rikhuijzer commented on base/errorshow.jl:
```diff
@@ -555,6 +560,7 @@ function show_method_candidates(io::IO, ex::MethodError, @nospecialize kwargs=()
                 i += 1
                 print(io, String(take!(line[1])))
             end
+            println(io) # extra newline for spacing to stacktrace
```
```suggestion
```

Seelengrab replied:
Where should the extra newline for an empty line go, if not here? The stacktrace printing happens in a completely different part of the stack, it doesn't even know about being in a thrown exception, I don't think.

rikhuijzer replied:
I thought that this was relating to the same thing as my previous comment. I'll click resolve

rikhuijzer reviewed (commented):
In response to your request for comments on Zulip

I like how you spotted and fixed the inconsistency between using `@` and `at` in the stacktrace versus the error message.

Seelengrab commented:
I was made aware of https://github.com/JuliaLang/julia/pull/40251 and now I'm thinking about unifying the two, since they're (in part) about the same exact thing: making methods & their uses easier to read when displayed.

Seelengrab commented:
Alright, I incorporated #40251 ([with permission](https://julialang.zulipchat.com/#narrow/stream/137791-general/topic/.60MethodError.60.20prettification/near/281091407)) into this PR, since that one kinda stalled and this one also touches method printing (though at a different place). I'll update the images at the top post.

Seelengrab commented:
Ah of course, doctests - those will have to wait for tomorrow.

Seelengrab commented:
While fixing the failing tests, I noticed that the ambiguity errors have their own code path here - so I'm changing those too.

JeffBezanson commented on base/methodshow.jl:
```diff
@@ -206,34 +205,49 @@ function sym_to_string(sym)
+            print(io, "; ")
+            for kw in kwargs
+                skw = sym_to_string(kw)
+                if QuoteNode(kw) in m.roots # then it's required
```
This is not a good way to test for this.

JeffBezanson replied:
IIUC, this just attempts to bold "required" keyword arguments? Since we don't really have a notion of required keyword arguments (just kwargs whose default is an error) I don't think this is essential and can just be left out.

Seelengrab replied:
Jameson had the [exact same objection](https://github.com/JuliaLang/julia/pull/40251#discussion_r788208317) in the original PR by @mcabbott :sweat_smile:

Seelengrab replied:
Yeah, that's what the code is supposed to do - and testing that explicitly makes it obvious that I didn't hit the last case `f(1,2,3; g=4)` with an extra keyword arg. Seems to be untested as well, so I'll add that test as well..

Seelengrab replied:
Turns out I just forgot to be on the correct branch when testing that - the error is nicely formatted already.

JeffBezanson commented on base/methodshow.jl:
```diff
@@ -253,18 +267,20 @@ function show_method_list_header(io::IO, ms::MethodList, namefmt::Function)
+    !iszero(n) && print(io, ":")
 end
 
+const METHODLIST_MODULECOLORS = [:cyan, :green, :yellow]
```
Why does this have a separate and different list of colors than stack traces?

Seelengrab replied:
I'll unify the colors between these and the stack traces :+1:

Seelengrab committed:
Make colors of printed modules consistent between stacktraces, methodlists and MethodErrors.

Remove bad detection of "required" keyword arguments

Fix missing line information for methods without a source file

Fix InteractiveUtils tests

Seelengrab committed:
Add NEWS.md entry

Fix last remaining reflection test

Seelengrab commented:
Alright, the only thing that's missing now should be the doctests (for real this time). I've also added a NEWS.md entry and rebased onto current master.

I'm still not 100% sure how to fix the doctests best, without someone in the future running into having to specifically exclude the non-existence of `[...]` everywhere :thinking:

Seelengrab committed:
Fix doctests

Seelengrab commented:
Alright, I think the doctests are fixed now, so this should be good to go once CI agrees. I took the liberty to squash the commits already, to keep history cleaner.

oscardssmith commented:
LGTM

Seelengrab committed:
Fix inconsistency of module numbers (again)

Seelengrab commented:
I missed an inconsistency with colors :| Here's what it looks like now:

![screenshot_2022-06-23_20-38-30_948777202](https://user-images.githubusercontent.com/11753998/175373882-30dce7c5-2cd4-4820-8b5e-621c528aab03.png)

Should be perfectly consistent now.

Seelengrab commented:
PowerPC failure is unrelated:

> JIT session error: Symbols not found: [ __gnu_f2h_ieee, __gnu_h2f_ieee ]
>  | Failure value returned from cantFail wrapped call
>  | Failed to materialize symbols: { (JuliaOJIT, { jfptr_Float16_2904, julia_Float16_2903 }) }
>  | UNREACHABLE executed at /cache/build/default-power8bot2-0/julialang/julia-master/usr/include/llvm/Support/Error.h:782!

Seelengrab commented:
Apologies for the force pushes after you added the `mergeme` label, fixing the colors broke some tests that should be fixed now (hopefully)..

Seelengrab commented:
The Apple Darwin failure seems unrelated:

>     nested task error: rmprocs: pids [5] not terminated after 30 seconds.

as is the PowerPC build failure:

> JIT session error: Symbols not found: [ __gnu_f2h_ieee, __gnu_h2f_ieee ]

No idea what's going on with the `Profile` failure in the "Allow Fail" category. Seems unrelated?

tecosaur commented:
I have to say, I'm not a fan of the choice to have variable names be greyed out. To quote past me in #40251 

> I think argument names deserve more emphasis, in my PR I use the magenta colour for them. for example there's the `join` method with signature `join(strings, delim, last)` and I think it's worth applying emphasising argument names for cases such as those.

Seelengrab commented:
The goal here was to make `MethodError`, `methods(f)` and stacktraces consistent in color - since it didn't seem like there was consensus about adding more colors there (and some said they dislike more color due to it adding more noise), I left the choice for later.

Perhaps this could be done in a future "customize all the colors" REPL theming PR, like suggested in #41435?
